polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -95,16 +97,28 @@ LogicalResult getTileSizes(const SmallVector<T> &nestedLoops,
 LogicalResult tile(MutableArrayRef<affine::AffineForOp> nestedLoops,
                    ArrayRef<Value> tileSizes,
                    SmallVectorImpl<affine::AffineForOp> &tiledNest) {
-  return tilePerfectlyNestedParametric(nestedLoops, tileSizes, &tiledNest);
+  SmallVector<affine::AffineForOp> newNestedLoops;
+  unsigned numLoops = nestedLoops.size();
+  LogicalResult res =
+      tilePerfectlyNestedParametric(nestedLoops, tileSizes, &newNestedLoops);
+  tiledNest = SmallVector<affine::AffineForOp>(
+      newNestedLoops.begin() + numLoops, newNestedLoops.end());
+  return res;
 }
 LogicalResult tile(SmallVectorImpl<scf::ForOp> &nestedLoops,
                    ArrayRef<Value> tileSizes,
                    SmallVectorImpl<scf::ForOp> &tiledNest) {
-  tile(nestedLoops, tileSizes, nestedLoops.back());
-  tiledNest = nestedLoops;
+  tiledNest = tile(nestedLoops, tileSizes, nestedLoops.back());
   return success();
 }
 
+void createLocalBarrier(OpBuilder builder) {
+  builder.create<spirv::ControlBarrierOp>(
+      builder.getUnknownLoc(), spirv::Scope::Workgroup, spirv::Scope::Workgroup,
```

Comment:
Can we actually guarantee the uniform control flow necessary for this operation? 

> When Execution is Workgroup or larger, behavior is undefined if this instruction is used in control flow that is non-uniform within Execution.

Reply:
I don't think we do. We could use `OpMemoryNamedBarrier` instead but I'm unsure if it is supported by our targets.

But otherwise, I don't think we need this op. I assume you are adding this to wait at the end of the memory copy. But to do this operation you have https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupAsyncCopy and the associated wait `OpGroupWaitEvents`. They both also require uniformity however.

Not sure `affine.dma_start` / `affine.dma_wait` would map to these (and it doesn't seem to lead anywhere anyway).

Reply:
Thanks for the suggestions. What do you mean by `and it doesn't seem to lead anywhere anyway`?

Reply:
In that case, we might as well just generate the spirv runtime barrier call for now.

Reply:
I understand that to inject a control barrier we need to prove there is no thread divergence. I was thinking that in this PR we *assume* the actual program doesn't have divergence and focus on generating a call to the SYCL RT for a workgroup barrier (given that we cannot generate a `spirv::ControlBarrierOp` or a `gpu.barrier (is lowered to spirv::ControlBarrierOp) because we can't yet lower a `spirv::ControlBarrierOp` to LLVM IR. 

So here is what we can do:
   - inject a call to the SYCL RT for a workgroup barrier
   - test that the codegen works end-to-end 
   - continue with the rest of the transformation (identify mem. accesses that should be prefetched in SLM, ...) 
   - go back and write an analysis to prove that the loop we want to transform is executed by all the threads in a workgroup
   
P.S.: 
I was hoping MLIR had an analysis to classify code regions that are not divergent w.r.t threads. But I haven't found one yet.

Reply:
Added spirv to llvm lowering for ControlBarrierOp. Tested that by adding ControlBarrierOp in the code, it can still link and run successfully.

Reply:
As there is no mention of the GroupAsyncCopy in the SPIR-V dialect, the easiest way is to implement the lowering from SPIR-V to LLVM for ControlBarrier. @sommerlukas @Naghasan What do you guys think of the current approach in this PR?

Reply:
I'm ok with the approach to use the control barrier now and add the necessary analysis of uniform execution later on, as long as we do not enable this pass by default before that is complete. 

Long term, adding and using `GroupAsycnCopy` might have some advantages, as some GPUs now have specialized hardware for that kind of operation, e.g., in context with AMX/Tensorcores.

Reply:
> I'm ok with the approach to use the control barrier now and add the necessary analysis of uniform execution later on, as long as we do not enable this pass by default before that is complete.

Great! We are not going to enable this pass by default before its completion. 
> 
> Long term, adding and using `GroupAsycnCopy` might have some advantages, as some GPUs now have specialized hardware for that kind of operation, e.g., in context with AMX/Tensorcores.

Agree. The current approach is to simulate `nd_item.barrier`, which is lowered to spirv ControlBarrier.